examples/client_execute_parallel_servers.pl

```diff
@@ -0,0 +1,128 @@
+#!/usr/bin/perl
+
+#
+# This script simulates parallel execution of commands
+# on several session to potentially different servers
+#
+
+use strict;
+use warnings;
+use Libssh::Session qw(:all);
+use Time::HiRes qw (usleep);
+
+my $ssh_host = "127.0.0.1";
+my $ssh_port = 22;
+my $ssh_user = "libssh";
+my $ssh_pass = "centreon";
+
+my $NUM_PARALLEL_CONNECTIONS = 3;
+
+sub print_output_callback {
+    my (%options) = @_;
+
+    if ($options{exit} == SSH_ERROR) {
+        my $error_msg = $options{error_msg};
+        print "== execution failed: $error_msg\n";
+        return 0;
+    }
+
+    my $rc = $options{exit_code};
+    my $stdout = $options{stdout};
+    my $stderr = $options{stderr};
+    print "=> rc = $rc\n";
+    print "=> stdout:\n$stdout";
+    print "=> stderr:\n$stderr" if (defined($stderr) && $stderr);
+    print '+'x8 . "\n";
+    return 1;
+}
+
+sub init_session {
+    my $session = Libssh::Session->new();
+    if ($session->options(host => $ssh_host, port => $ssh_port, user => $ssh_user) != SSH_OK) {
+        print $session->error() . "\n";
+        exit(1);
+    }
+
+    if ($session->connect() != SSH_OK) {
+        print $session->error() . "\n";
+        exit(1);
+    }
+
+    if ($session->auth_publickey_auto() != SSH_AUTH_SUCCESS) {
+        if ($session->auth_password(password => $ssh_pass) != SSH_AUTH_SUCCESS) {
+            printf("auth issue: %s\n", $session->error(GetErrorSession => 1));
+            exit(1);
+        }
+    }
+
+    return $session;
+}
+
+sub init_nsessions {
+    my ($num) = @_;
+    die "You can't init less than one session." if ($num < 1);
+
+    my @all_sessions = ();
+    for (0 .. $num - 1) {
+        my $session = init_session();
+        push(@all_sessions, $session);
+    }
+    return @all_sessions;
+}
+
+my @all_sessions = init_nsessions($NUM_PARALLEL_CONNECTIONS);
+my @all_channels = ();
+my %has_returned = ();
+
+my $sleep_seconds = 1;
+for my $i (0 .. $#all_sessions) {
+    my $session = $all_sessions[$i];
+    my $cmd = "sleep $sleep_seconds && ls ~";
+    print "== calling non-blocking '$cmd' for session[$i]\n";
+    my $channel_id = $session->add_command_internal(
+        command => {cmd => $cmd, callback => \&print_output_callback}
+    );
```

Comment:
Here add_command_internal is used outside of Session.pm which is probably not great syntax wise but it works great for my use case. Any ideas on how to better leverage the existing logic for opening a channel, starting a command on, writing input_data to it, and attaching a callback will be greatly appreciated.